{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "private_outputs": true,
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/badonyi/mechanism-prediction/blob/main/mechanism_prediction.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4hDavFHmGneJ",
        "cellView": "form",
        "collapsed": true
      },
      "outputs": [],
      "source": [
        "# Setup ------------------------------------------------------------------------\n",
        "# prevent false objects in subsequent runs, but keep cached data\n",
        "rm(list = ls()[!ls() %in% c('ref_set', 'priors', 'fx_raw', 'handle')])\n",
        "\n",
        "# ANSI escape codes for text colour\n",
        "RED <- '\\033[31m'\n",
        "GREEN <- '\\033[32m'\n",
        "RESET <- '\\033[0m'\n",
        "\n",
        "# canonical amino acid vectors\n",
        "aa1 <- strsplit('ARNDCEQGHILKMFPSTWYV', '')[[1]]\n",
        "aa3 <- c('ALA', 'ARG', 'ASN', 'ASP', 'CYS', 'GLU', 'GLN', 'GLY', 'HIS', 'ILE',\n",
        "         'LEU', 'LYS', 'MET', 'PHE', 'PRO', 'SER', 'THR', 'TRP', 'TYR', 'VAL')\n",
        "\n",
        "# UniProt REST API and AlphaFold database\n",
        "up_api <- paste0('https://rest.uniprot.org/uniprotkb/search?query=',\n",
        "                 'reviewed:true+AND+%s+organism_id:9606&format=tsv')\n",
        "af_db <- 'https://alphafold.ebi.ac.uk/files/AF-%s-F1-model_v6.pdb'\n",
        "\n",
        "# helper functions\n",
        "density_sj <- function(x) density(x, bw = 'SJ', adjust = 3, n = 1024)\n",
        "bayes <- function(p, prior) (p * prior) / (p * prior + (1 - p) * (1 - prior))\n",
        "redstop <- function(x) stop(RED, x, RESET, call. = FALSE)\n",
        "\n",
        "\n",
        "# User options -----------------------------------------------------------------\n",
        "#@markdown ### <b><font color='#85b0f5'>Specify a human gene</font></b>\n",
        "#@markdown #### Gene name\n",
        "gene = 'SCN9A' #@param {type:\"string\"}\n",
        "gene <- trimws(gene)\n",
        "\n",
        "#@markdown #### <b>or</b> UniProt accession number\n",
        "id = '' #@param {type:\"string\"}\n",
        "id <- toupper(trimws(id))\n",
        "\n",
        "#@markdown ### <b><font color='#85b0f5'>Input variants</font></b>\n",
        "#@markdown Example: `A108S, L730H, T454P`\n",
        "variants = 'F216S,I234T,N395K,V400M,L834R,I859T,L869H,L869F,Q886E,R907W,Y1001C,V1002L,V1309D,V1309F,V1310F,F1460V,I1472N,I1472T,F1473V,T1475I,L1623P,A1643E,T660M,G867R,R907Q,A1143P,V1310I,C1350R,D1959A,G867D,M1639K' #@param {type:\"string\"}\n",
        "\n",
        "\n",
        "# Check 1 - Input variants -----------------------------------------------------\n",
        "dr_split <- toupper(trimws(strsplit(variants, ',')[[1]]))\n",
        "\n",
        "# wild-type, mutant, and residue positions; used later to match input\n",
        "wt <- substr(dr_split, 1, 1)\n",
        "mut <- substr(dr_split, nchar(dr_split), nchar(dr_split))\n",
        "dr <- as.integer(gsub('[^0-9]', '', dr_split))\n",
        "\n",
        "if (length(dr) < 3)\n",
        "  redstop('EDC calculation requires at least 3 residue positions.')\n",
        "\n",
        "if (!all(mut %in% aa1))\n",
        "  redstop('Missing or non-canonical amino acids are not supported.')\n",
        "\n",
        "\n",
        "# Check 2 - UniProt accession or gene symbol -----------------------------------\n",
        "if (gene == '' & id == '')\n",
        "  redstop('Please provide either a UniProt ID or a gene name.')\n",
        "\n",
        "if (gene != '') {\n",
        "  # match gene symbol with UniProt accession via UniProt REST API\n",
        "  up_tbl <- read.delim(sprintf(up_api, gene), sep = '\\t')\n",
        "  if (nrow(up_tbl) > 0) {\n",
        "    up_lst <- strsplit(up_tbl$Gene.Names, ' ')\n",
        "    up_lst <- up_lst[lengths(up_lst) > 0]\n",
        "    up_idx <- sapply(up_lst, function(x) any(tolower(x) == tolower(gene)))\n",
        "    id <- up_tbl[up_idx, ]$Entry[1]\n",
        "  } else {\n",
        "    redstop('Gene could not be mapped to a UniProt accession ID.')\n",
        "  }\n",
        "}\n",
        "\n",
        "upreg <- '[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9]([A-Z][A-Z0-9]{2}[0-9]){1,2}'\n",
        "if (!grepl(upreg, id))\n",
        "  redstop('UniProt accession not supported.')\n",
        "\n",
        "\n",
        "# Check 3 - Fetch molecular mechanism predictions ------------------------------\n",
        "if (!exists('priors')) {\n",
        "  download.file('https://osf.io/download/e6b9c/', destfile = 'priors.tsv')\n",
        "  priors <- read.delim('priors.tsv', sep = '\\t')\n",
        "}\n",
        "\n",
        "if (id %in% priors$uniprot_id) {\n",
        "  prior_set <- priors[priors$uniprot_id == id, ]\n",
        "  pLOF <- prior_set$pLOF\n",
        "  pDN  <- prior_set$pDN\n",
        "  pGOF <- prior_set$pGOF\n",
        "} else {\n",
        "  redstop('UniProt accession not supported.')\n",
        "}\n",
        "\n",
        "\n",
        "# Fetch AlphaFold predicted model ----------------------------------------------\n",
        "cat('INFO     Using the AlphaFold model',\n",
        "    paste0('(', sprintf(af_db, id), ')'))\n",
        "\n",
        "tryCatch({\n",
        "  # attempt to parse ATOM lines\n",
        "  a <- grep('^ATOM', readLines(sprintf(af_db, id), warn = FALSE), value = TRUE)\n",
        "}, warning = function(w) {\n",
        "  redstop('Structure may not exist for the given UniProt ID.')\n",
        "})\n",
        "\n",
        "\n",
        "# Parse alpha carbon coordinates and pLDDT from b-factor column ----------------\n",
        "a <- a[trimws(substr(a, 14, 15)) == 'CA']\n",
        "xyz <- data.frame(\n",
        "  r = as.integer(substr(a, 23, 26)),\n",
        "  x = as.numeric(substr(a, 31, 38)),\n",
        "  y = as.numeric(substr(a, 39, 46)),\n",
        "  z = as.numeric(substr(a, 47, 54)),\n",
        "  b = as.numeric(substr(a, 61, 66))\n",
        ")\n",
        "\n",
        "\n",
        "# Extract sequence from structure ----------------------------------------------\n",
        "uniqres <- trimws(substr(a, 18, 21))\n",
        "pdb_seq <- setNames(aa1, aa3)[uniqres]\n",
        "names(pdb_seq) <- xyz$r # maps relative to PDB numbering\n",
        "wt_in_sturcture <- sapply(dr, function(i) pdb_seq[i])\n",
        "\n",
        "\n",
        "# Process sturcture for input residue context ----------------------------------\n",
        "dr_remaining <- xyz[xyz$r %in% dr, ]$r\n",
        "xyz_missing <- sort(unique(dr[!dr %in% dr_remaining]))\n",
        "\n",
        "if (length(xyz_missing) > 0)\n",
        "  cat(RED, '\\nWARNING  Residue(s) not found in structure:',\n",
        "      paste(xyz_missing, collapse = ', '), RESET)\n",
        "\n",
        "# filter above the pLDDT threshold\n",
        "xyz <- xyz[xyz$b >= 70, ]\n",
        "\n",
        "# disease and non-disease residue subsets\n",
        "dc <- xyz[ xyz$r %in% dr, c('x','y','z')]\n",
        "ac <- xyz[!xyz$r %in% dr, c('x','y','z')]\n",
        "\n",
        "# residues removed by the pLDDT filter\n",
        "dr_remaining <- xyz[xyz$r %in% dr, ]$r\n",
        "n_res <- length(dr_remaining) # required for mLOF\n",
        "plddt_remove <- sort(unique(dr[!dr %in% dr_remaining]))\n",
        "plddt_remove <- plddt_remove[!plddt_remove %in% xyz_missing]\n",
        "\n",
        "if (length(plddt_remove) > 0)\n",
        "  cat(RED, '\\nWARNING  Residue(s)', paste(plddt_remove, collapse = ', '),\n",
        "      'are below the pLDDT cutoff for EDC calculation.', RESET)\n",
        "\n",
        "if (length(dr_remaining) < 3)\n",
        "  redstop('Not enough residues to calculate EDC.')\n",
        "\n",
        "\n",
        "# Report any mismatched amino acids --------------------------------------------\n",
        "aa_mismatch <- data.frame(\n",
        "  input = as.character(dr_split[wt != wt_in_sturcture]),\n",
        "  structure = as.character(wt_in_sturcture[wt != wt_in_sturcture])\n",
        ")\n",
        "aa_mismatch <- aa_mismatch[order(dr), ]\n",
        "aa_mismatch <- aa_mismatch[!is.na(aa_mismatch$input), ]\n",
        "\n",
        "if (nrow(aa_mismatch) > 0) {\n",
        "  cat(RED, '\\nWARNING  Residue(s) in structure different from input:\\n\\n')\n",
        "  print(aa_mismatch, row.names = FALSE)\n",
        "  cat(RESET)\n",
        "}\n",
        "\n",
        "\n",
        "# Calculate EDC ----------------------------------------------------------------\n",
        "cat('\\nINFO     Calculating EDC... ')\n",
        "a_mat <- {\n",
        "  # pairwise distances between dc and ac using matrix arithmetics\n",
        "  a <- as.matrix(ac)\n",
        "  b <- as.matrix(dc)\n",
        "  an <- apply(a, 1, function(x) crossprod(x, x))\n",
        "  bn <- apply(b, 1, function(x) crossprod(x, x))\n",
        "  nn <- outer(an, bn, '+')\n",
        "  sqrt(nn - 2 * tcrossprod(a, b))\n",
        "}\n",
        "\n",
        "# minimum distances for non-disease subset\n",
        "a_min <- apply(a_mat, 1, min)\n",
        "\n",
        "# pairwise and minimum distances for disease subset\n",
        "d_mat <- as.matrix(dist(dc))\n",
        "d_min <- apply(d_mat, 1, function(row) min(row[row != 0]))\n",
        "\n",
        "# EDC value\n",
        "edc <- mean(log(a_min)) / mean(log(d_min))\n",
        "\n",
        "if (is.na(edc))\n",
        "  redstop('EDC could not be calculated.')\n",
        "\n",
        "cat('DONE')\n",
        "\n",
        "\n",
        "# Fetch ΔΔG ranks --------------------------------------------------------------\n",
        "cat('\\nINFO     Retrieving ΔΔGrank scores... ')\n",
        "if (!exists('fx_raw') || handle != priors[priors$uniprot_id == id, ]$handle) {\n",
        "  # get OSF handle for UniProt ID and download file\n",
        "  handle <- priors[priors$uniprot_id == id, ]$handle\n",
        "  download.file(paste0('https://osf.io/download/', handle), destfile = 'fx.tsv')\n",
        "  fx_raw <- read.delim(file = 'fx.tsv', sep = '\\t')\n",
        "}\n",
        "cat('DONE')\n",
        "\n",
        "# parse out ΔΔG rank values into a named vector\n",
        "fx <- fx_raw[fx_raw$uniprot_id == id, ]$ddg_rank\n",
        "fx <- strsplit(fx, '\\\\|')[[1]]\n",
        "vars <- as.numeric(sub('^[^,]*,', '', fx))\n",
        "names(vars) <- sub(',.*', '', fx)\n",
        "\n",
        "# compute ΔΔG rank for valid input variants\n",
        "ddg_rank <- vars[paste0(wt, dr, mut)]\n",
        "n_var <- length(ddg_rank) # required for mLOF\n",
        "ddg_rank <- mean(vars[paste0(wt, dr, mut)], na.rm = TRUE)\n",
        "\n",
        "if (is.na(ddg_rank))\n",
        "  redstop('ΔΔGrank could not be calculated.')\n",
        "\n",
        "\n",
        "# Compute mLOF -----------------------------------------------------------------\n",
        "if (!exists('ref_set')) {\n",
        "  download.file('https://osf.io/download/7vgh4/', destfile = 'ref_set.tsv')\n",
        "  ref_set <- read.delim('ref_set.tsv', sep = '\\t')\n",
        "}\n",
        "\n",
        "# empirical EDC and ΔΔG rank distributions for LOF and non-LOF genes\n",
        "lof_edc <- ref_set[ref_set$mechanism == 'LOF', ]$edc\n",
        "non_lof_edc <- ref_set[ref_set$mechanism == 'non-LOF', ]$edc\n",
        "lof_ddg <- ref_set[ref_set$mechanism == 'LOF', ]$ddg_rank\n",
        "non_lof_ddg <- ref_set[ref_set$mechanism == 'non-LOF', ]$ddg_rank\n",
        "\n",
        "# kernel density estimates\n",
        "lof_edc_d <- density_sj(lof_edc)\n",
        "non_lof_edc_d <- density_sj(non_lof_edc)\n",
        "lof_ddg_d <- density_sj(lof_ddg)\n",
        "non_lof_ddg_d <- density_sj(non_lof_ddg)\n",
        "\n",
        "# probability caps\n",
        "lof_edc_cap <- quantile(lof_edc, 0.1)\n",
        "non_lof_edc_cap <- quantile(non_lof_edc, 0.9)\n",
        "lof_ddg_cap <- quantile(lof_ddg, 0.9)\n",
        "nonlof_ddg_cap <- quantile(non_lof_ddg, 0.1)\n",
        "\n",
        "# marginal probabilities\n",
        "p_lof_edc <- {\n",
        "  # cap scenario\n",
        "  capped_edc <- edc\n",
        "  capped_edc[capped_edc > non_lof_edc_cap] <- non_lof_edc_cap\n",
        "  capped_edc[capped_edc < lof_edc_cap] <- lof_edc_cap\n",
        "\n",
        "  # indices of minimum values\n",
        "  lof_idx <- sapply(\n",
        "    X = capped_edc,\n",
        "    FUN = function(observation) {\n",
        "      which.min(abs(lof_edc_d$x - observation))\n",
        "    }\n",
        "  )\n",
        "\n",
        "  nonlof_idx <- sapply(\n",
        "    X = capped_edc,\n",
        "    FUN = function(observation) {\n",
        "      which.min(abs(non_lof_edc_d$x - observation))\n",
        "    }\n",
        "  )\n",
        "\n",
        "  # density values corresponding to those indices\n",
        "  density_value_nonlof <- non_lof_edc_d$y[nonlof_idx]\n",
        "  density_value_lof <- lof_edc_d$y[lof_idx]\n",
        "\n",
        "  # LOF (EDC) probability\n",
        "  density_value_lof / (density_value_lof + density_value_nonlof)\n",
        "}\n",
        "\n",
        "p_lof_ddg <- {\n",
        "  # cap scenario\n",
        "  capped_ddg_rank <- ddg_rank\n",
        "  capped_ddg_rank[capped_ddg_rank < nonlof_ddg_cap] <- nonlof_ddg_cap\n",
        "  capped_ddg_rank[capped_ddg_rank > lof_ddg_cap] <- lof_ddg_cap\n",
        "\n",
        "  # indices of minimum values\n",
        "  lof_idx <- sapply(\n",
        "    X = capped_ddg_rank,\n",
        "    FUN = function(observation) {\n",
        "      which.min(abs(lof_ddg_d$x - observation))\n",
        "    }\n",
        "  )\n",
        "\n",
        "  nonlof_idx <- sapply(\n",
        "    X = capped_ddg_rank,\n",
        "    FUN = function(observation) {\n",
        "      which.min(abs(non_lof_ddg_d$x - observation))\n",
        "    }\n",
        "  )\n",
        "\n",
        "  # density values corresponding to those indices\n",
        "  density_value_nonlof <- non_lof_ddg_d$y[nonlof_idx]\n",
        "  density_value_lof <- lof_ddg_d$y[lof_idx]\n",
        "\n",
        "  # LOF (ddG) probability\n",
        "  density_value_lof / (density_value_lof + density_value_nonlof)\n",
        "}\n",
        "\n",
        "\n",
        "# missense LOF likelihood (unadjusted)\n",
        "mLOF <- (p_lof_edc * n_var + p_lof_ddg * n_res) / (n_var + n_res)\n",
        "\n",
        "\n",
        "# Calculate mLOF and posterior likelihoods -------------------------------------\n",
        "mLOF <- (p_lof_edc * n_var + p_lof_ddg * n_res) / (n_var + n_res)\n",
        "postDN <- bayes(p = 1 - mLOF, prior = pDN)\n",
        "postGOF <- bayes(p = 1 - mLOF, prior = pGOF)\n",
        "postLOF <- bayes(p = mLOF, prior = pLOF)\n",
        "\n",
        "\n",
        "# Print metric tables ----------------------------------------------------------\n",
        "knitr::kable(\n",
        "  x = c(\n",
        "    'EDC' = edc,\n",
        "    'ΔΔGrank' = ddg_rank,\n",
        "    'mLOF' = mLOF\n",
        "  ),\n",
        "  digits = 3,\n",
        "  col.names = c('Structure-based metrics', ''),\n",
        "  format = 'simple'\n",
        ")\n",
        "\n",
        "knitr::kable(\n",
        "  x = c(\n",
        "    'pLOF' = pLOF,\n",
        "    'pGOF' = pGOF,\n",
        "    'pDN' = pDN\n",
        "  ),\n",
        "  digits = 3,\n",
        "  col.names = c('Prior mechanism propensities', ''),\n",
        "  format = 'simple'\n",
        ")\n",
        "\n",
        "knitr::kable(\n",
        "  x = sort(\n",
        "    c(\n",
        "      'postDN' = postDN,\n",
        "      'postGOF' = postGOF,\n",
        "      'postLOF' = postLOF\n",
        "    ),\n",
        "    decreasing = TRUE\n",
        "  ),\n",
        "  digits = 3,\n",
        "  col.names = c('Ranked mechanism-specific posteriors', ''),\n",
        "  format = 'simple'\n",
        ")\n",
        "\n",
        "\n",
        "# Provide a verdict ------------------------------------------------------------\n",
        "suggestion <- c(\n",
        "  'dominant-negative', 'gain-of-function', 'loss-of-function'\n",
        ")[which(c(postDN, postGOF, postLOF) == pmax(postDN, postGOF, postLOF))][1]\n",
        "\n",
        "if (suggestion != 'loss-of-function' & mLOF > 0.508) {\n",
        "  cat('\\n\\nVERDICT', GREEN,\n",
        "      'While the structural properties of the variants are more consistent',\n",
        "      'with a LOF effect,\\n         based upon the',\n",
        "      'properties of the gene, a', suggestion, 'mechanism is expected.', RESET)\n",
        "} else if (suggestion == 'loss-of-function' & mLOF < 0.508) {\n",
        "  cat('\\n\\nVERDICT', GREEN,\n",
        "      'While the structural properties of the variants are more consistent',\n",
        "      'with a non-LOF effect,\\n         based upon the',\n",
        "      'properties of the gene, a', suggestion, 'mechanism is expected.', RESET)\n",
        "} else {\n",
        "  cat('\\n\\nVERDICT', GREEN,\n",
        "      'A', suggestion, 'mechanism is most likely for the variants.', RESET)\n",
        "}\n",
        "cat('\\n\\n')\n",
        "\n",
        "# Generate plots ---------------------------------------------------------------\n",
        "options(repr.plot.width = 16, repr.plot.height = 4.5)\n",
        "par(mfrow = c(1, 3), mar = c(5, 4, 4, 2) + 0.25)\n",
        "\n",
        "# mechanism prior plot\n",
        "prior_col <- c('dominant-negative' = 'pDN',\n",
        "               'gain-of-function' = 'pGOF',\n",
        "               'loss-of-function' = 'pLOF')[suggestion]\n",
        "all_priors <- sort(priors[[prior_col]])\n",
        "obs_prior <- mget(prior_col)\n",
        "prior_rnk <- which(all_priors == obs_prior)[1]\n",
        "prior_pctl <- sprintf('%.1fth percentile', prior_rnk / length(all_priors) * 100)\n",
        "\n",
        "plot(all_priors,\n",
        "     main = 'Observed prior mechanism propensity\\nfor the suggested mechanism',\n",
        "     type = 'l',\n",
        "     xlab = 'All human genes',\n",
        "     ylab = prior_col,\n",
        "     ylim = c(0, 1),\n",
        "     col = 'grey50',\n",
        "     lwd = 3,\n",
        "     cex.main = 1.8,\n",
        "     cex.lab = 1.8,\n",
        "     cex.axis = 1.8)\n",
        "points(x = prior_rnk,\n",
        "       y = obs_prior,\n",
        "       col = 'firebrick',\n",
        "       pch = 19,\n",
        "       cex = 2)\n",
        "text(x = 7000,\n",
        "     y = 1,\n",
        "     labels = prior_pctl,\n",
        "     col = 'firebrick',\n",
        "     cex = 1.75,\n",
        "     adj = c(0, 1))\n",
        "\n",
        "# EDC density plot\n",
        "plot(lof_edc_d,\n",
        "     main = 'Probability density of EDC',\n",
        "     xlab = 'EDC',\n",
        "     ylab = 'Density',\n",
        "     col = '#fb8072',\n",
        "     lwd = 3,\n",
        "     xlim = c(0.5, 2.5),\n",
        "     ylim = c(0, 2),\n",
        "     cex.main = 1.8,\n",
        "     cex.lab = 1.8,\n",
        "     cex.axis = 1.8)\n",
        "lines(non_lof_edc_d, lwd = 3, col = '#80b1d3')\n",
        "abline(v = edc, col = '#1A1A1A80', lwd = 2, lty = 2)\n",
        "legend(x = 1.5, y = 2,\n",
        "       legend = c('LOF', 'non-LOF'),\n",
        "       col = c('#fb8072', '#80b1d3'),\n",
        "       lwd = 3,\n",
        "       lty = 1,\n",
        "       cex = 1.75,\n",
        "       seg.len = 0.8,\n",
        "       x.intersp = 0.5,\n",
        "       y.intersp = 1.5,\n",
        "       bty = 'n')\n",
        "\n",
        "# ΔΔG rank density plot\n",
        "plot(lof_ddg_d,\n",
        "     main = 'Probability density of ΔΔGrank',\n",
        "     xlab = bquote(ΔΔG[rank]),\n",
        "     ylab = 'Density',\n",
        "     col = '#fb8072',\n",
        "     lwd = 3,\n",
        "     xlim = c(0, 1),\n",
        "     ylim = c(0, 3),\n",
        "     cex.main = 1.8,\n",
        "     cex.lab = 1.8,\n",
        "     cex.axis = 1.8)\n",
        "lines(non_lof_ddg_d, lwd = 3, col = '#80b1d3')\n",
        "abline(v = ddg_rank, col = '#1A1A1A80', lwd = 2, lty = 2)\n",
        "legend(x = -0.1, y = 3,\n",
        "       legend = c('LOF', 'non-LOF'),\n",
        "       col = c('#fb8072', '#80b1d3'),\n",
        "       lwd = 3,\n",
        "       lty = 1,\n",
        "       cex = 1.75,\n",
        "       seg.len = 0.8,\n",
        "       x.intersp = 0.5,\n",
        "       y.intersp = 1.5,\n",
        "       bty = 'n')"
      ]
    }
  ]
}